concrete/src/Entity/File/Version.php

```diff
@@ -657,697 +872,862 @@ public function getVersionLogComments()
+        $mimeType = Util::guessMimeType($fileName, $fileContents);
+        $filesystem->write(
+            $path,
+            $fileContents,
+            [
+                'visibility' => AdapterInterface::VISIBILITY_PUBLIC,
+                'mimetype' => $mimeType,
+            ]
+        );
+        $this->updateFile($fileName, $prefix);
     }
 
     /**
-     * Returns a url that can be used to download a file, will force the download of all file types, even if your browser can display them.
+     * Delete this version of the file.
      *
-     * @return string
+     * @param bool $deleteFilesAndThumbnails should we delete the actual file and the thumbnails?
      */
-    public function getForceDownloadURL()
+    public function delete($deleteFilesAndThumbnails = false)
     {
-        $c = Page::getCurrentPage();
-        $cID = ($c instanceof Page) ? $c->getCollectionID() : 0;
+        $app = Application::getFacadeApplication();
+        $db = $app->make(Connection::class);
+        $em = $app->make(EntityManagerInterface::class);
+        $category = $this->getObjectAttributeCategory();
 
-        return View::url('/download_file', 'force', $this->getFileID(), $cID);
+        foreach ($this->getAttributes() as $attribute) {
+            $category->deleteValue($attribute);
+        }
+
+        $db->executeQuery('DELETE FROM FileVersionLog WHERE fID = ? AND fvID = ?', [$this->getFileID(), $this->fvID]);
+
+        if ($deleteFilesAndThumbnails) {
+            $types = Type::getVersionList();
+            foreach ($types as $type) {
+                $this->deleteThumbnail($type);
+            }
+            try {
+                $fsl = $this->getFile()->getFileStorageLocationObject()->getFileSystemObject();
+                $fre = $this->getFileResource();
+                if ($fsl->has($fre->getPath())) {
+                    $fsl->delete($fre->getPath());
+                }
+            } catch (FileNotFoundException $e) {
+            }
+        }
+        $em->remove($this);
+        $em->flush();
     }
 
     /**
-     * Forces the download of a file.
+     * Get an abstract object to work with the actual file resource (note: this is NOT a concrete5 File object).
+     *
+     * @throws \League\Flysystem\FileNotFoundException
+     *
+     * @return \League\Flysystem\File
      */
-    public function forceDownload()
+    public function getFileResource()
     {
-        session_write_close();
-        $fre = $this->getFileResource();
-
+        $app = Application::getFacadeApplication();
+        $cf = $app->make('helper/concrete/file');
         $fs = $this->getFile()->getFileStorageLocationObject()->getFileSystemObject();
-        $response = new FlysystemFileResponse($fre->getPath(), $fs);
-
-        $response->setContentDisposition(ResponseHeaderBag::DISPOSITION_ATTACHMENT);
-        $response->prepare(\Request::getInstance());
+        $fo = $fs->get($cf->prefix($this->fvPrefix, $this->fvFilename));
```

Comment:
`return $fs->get($cf->prefix($this->fvPrefix, $this->fvFilename));`?

Reply:
I just kept the existing code, and IMHO it's better to keep different lines for different actions...